src/c3py/history.py

```diff
@@ -61,6 +61,20 @@ def __init__(self, data: dict[str, list[Operation]]):
         for process, ops in data.items():
             for i in range(len(ops) - 1):
                 self.poset.order_try(f"{process}.{i + 1}", f"{process}.{i + 2}")
+    
+    # co = (po U wr)^+
+    def causal_order(self):
+        ch = deepcopy(self)
+        for id1, op1 in self.label.items():
+            if op1.method == "wr":
+                continue
+
+            arg = op1.arg
+            ret = op1.ret
+            for id2, op2 in self.label.items():
+                if op2.method == "wr" and op2.arg == (arg, ret):
+                    ch.poset.link(id2, id1)
```

Comment:
It's my fault for not explicitly stating this, but currently, the graph (`G`) in the `Poset` class maintains two invariants (that comes from the definition of strong partial order). 

First, the graph must be acyclic. This is to keep the relation irreflexive.

Second, the graph must be transitive closure. That is, if there is a path from a to b and b to c, there must also be an edge from a to c. 

I'm fine with violating the property temporarily, but I have a feeling that we should NOT return an instance of `Poset` that is actually not a partial order.

I would probably change the function to `def make_co(self) -> Self | None` and return `None` if there is a cycle in the resulting poset. You probably don't need to check every time you add an edge and just need to check once at the end.

You also want to make sure the returned graph is transitive. You should be able to use a library function from `networkx`.